### ISSUE TITLE
Support tab completion of LaTeX-like abbreviations in the Raku REPL

### ISSUE THREAD

**Issue by ohmycloud** [rakudo/rakudo#3831]
# The Problem

Neither Raku's REPL nor [Comma](https://commaide.com)'s REPL support tab completion of LaTeX-like abbreviations. 
However, Julia support tab completion of LaTeX-like abbreviations in the Julia REPL very well. It's easy to [enter Unicode in Julia's REPL](https://docs.julialang.org/en/v1/manual/unicode-input/). And Ipython also support  [tab completion of LaTeX-like abbreviations](https://github.com/JuliaLang/julia/issues/32394#issuecomment-505251713).

## Expected Behavior

Make tab completion of LaTeX-like abbreviations in Raku REPL possible.

## Actual Behavior

Neither Raku's REPL nor [Comma](https://commaide.com)'s REPL support tab completion of LaTeX-like abbreviations.

**Comment by JJ** [comment]
The thing is that Rakudo provides only a very basic REPL, with most "good" functionality spread out to Readline and Linenoise. It would probably make more sense to add it to one of those, or create one that's based on them and provides this functionality.

**Comment by AlexDaniel** [comment]
This proposal looked weird to me at first, but then I realized how liked that feature is in other languages.

> with most "good" functionality spread out to Readline and Linenoise. It would probably make more sense to add it to one of those

What do you even [mean](https://github.com/rakudo/rakudo/blob/8134470538457c68164eb64f90e8d3e946d235a0/src/core.c/REPL.pm6)? Linenoise and Readline are just modules with bindings. Yes, Linenoise and Readline themselves provide some features, but the REPL implementation is in Rakudo, and it is just using the needed bits in order to not reinvent the wheel for basic stuff. Tab completion comes from Rakudo, and LaTeX-like abbreviations can be done there too. Or am I wrong?

> The thing is that Rakudo provides only a very basic REPL

I would've expressed it more bluntly but that works too.

**Comment by JJ** [comment]
Well, looking at python vs. bpython, for instance, bpython provides
completions, python doesn't. It's not a basic functionality, same as
history navigation is not, which is why it's provided by Linenoise.
However, LineNoise does provide completions, and it actually uses the
support of rakudo with the function completions-for-line.

Might be the case that functionality should be supported in the basic REPL;
for the time being, the election of providing support for it, but letting
REPL plugins deploy it seems right.

**Comment by ohmycloud** [comment]
Maybe a module(in the core or the third party) like [REPLCompletions.jl](https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/REPLCompletions.jl) can help. see also [Mapping from LaTeX math symbol to the corresponding Unicode codepoint](https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/latex_symbols.jl) 

And one feature I missed is `backspace` in the case of Unicode. In the Julia REPL, for example, when typed `1²3` and backspace, it will delete 3, ², 1. One by one. But in the Raku REPL,  the ² becomes `?` character. It seems character are not align. I don'tknow if we can refer to [edit_backspace](https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/LineEdit.jl#L793) in Julia REPL.

**Comment by lizmat** [comment]
> But in the Raku REPL, the ² becomes ? character. It seems character are not align. I don't know if we can refer to [edit_backspace](https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/LineEdit.jl#L793) in Julia REPL.

I cannot reproduce this with the 2024.10 release

**Comment by lizmat** [comment]
I've started development of the [`REPL`](https://raku.land/zef:lizmat/REPL) module to address some of the concerns raised in this issue.  It now supports completions (based on the Raku environment) **and** completions of Unicode code point names in unclosed `\c[` sequences.  E.g. entering `> say "\c[love` and pressing TAB will cycle between `LOVE HOTEL`, `LOVE LETTER` and `I LOVE YOU HAND SIGN`, the three Unicode code points that have the word "LOVE" in their name.

I'm still not sure what you mean by "LaTeX-like abbreviations", could you elaborate so I can consider integrating them into this REPL module?

**Comment by ohmycloud** [comment]
> > But in the Raku REPL, the ² becomes ? character. It seems character are not align. I don't know if we can refer to [edit_backspace](https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/LineEdit.jl#L793) in Julia REPL.
> 
> I cannot reproduce this with the 2024.10 release

yep, I'm using an older version of Rakudo(v2022.07), and  on another machine, I installed Rakudo 2024.05, it's ok to delete `1²3`, and thanks for the great work!

**Comment by ohmycloud** [comment]
> I've started development of the [`REPL`](https://raku.land/zef:lizmat/REPL) module to address some of the concerns raised in this issue. It now supports completions (based on the Raku environment) **and** completions of Unicode code point names in unclosed `\c[` sequences. E.g. entering `> say "\c[love` and pressing TAB will cycle between `LOVE HOTEL`, `LOVE LETTER` and `I LOVE YOU HAND SIGN`, the three Unicode code points that have the word "LOVE" in their name.
> 
> I'm still not sure what you mean by "LaTeX-like abbreviations", could you elaborate so I can consider integrating them into this REPL module?

The LaTeX-like abbreviations, for example, in the julia REPL, when I type `\`, `^`, and `2`, then press the `TAB`, it output superscript  `²`, which i think is very convenient, and I don't have to copy the superscript  somewhere.

```
julia> \^2 TAB  ²
julia> \_2 TAB  ₂
julia> \guillemotleft TAB «
julia> \guillemotright TAB »
```

**Comment by lizmat** [comment]
[REPL 0.0.9](https://raku.land/zef:lizmat/REPL?v=0.0.9) now implements "\_123" and "\^123" tab completions for subscript / superscript.  And tab completions with "\word" where word as taken as a word in a unicode codepoint name: at this point it doesn't feel right to take in all of the LaTeX abbreviations.

But of course, one is free to create a plugin that does so: the `REPL` module allows plugins with the `:additional-completions` argument to the `REPL` object instantiation.

**Comment by ohmycloud** [comment]
> FWIW, [`REPL 0.0.10`](https://raku.land/zef:lizmat/REPL) implements casing tab-completion: "foo!" will tab-complete to "FOO", "foo", "Foo" respectively, allowing you to fix those cases where you the casing wrong on a word.

It'a amazing work!
